internal/balancer/gracefulswitch/gracefulswitch_test.go

```diff
@@ -0,0 +1,1083 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package gracefulswitch
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/internal/grpcsync"
+	"google.golang.org/grpc/internal/grpctest"
+	"google.golang.org/grpc/internal/testutils"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/serviceconfig"
+)
+
+const (
+	defaultTestTimeout      = 5 * time.Second
+	defaultTestShortTimeout = 10 * time.Millisecond
+)
+
+type s struct {
+	grpctest.Tester
+}
+
+func Test(t *testing.T) {
+	grpctest.RunSubTests(t, s{})
+}
+
+func setup(t *testing.T) (*testutils.TestClientConn, *Balancer) {
+	tcc := testutils.NewTestClientConn(t)
+	return tcc, NewBalancer(tcc, balancer.BuildOptions{})
+}
+
+// TestSuccessfulFirstUpdate tests a basic scenario for the graceful switch load
+// balancer, where it is setup with a balancer which should populate the current
+// load balancer. Any ClientConn updates should then be forwarded to this
+// current load balancer.
+func (s) TestSuccessfulFirstUpdate(t *testing.T) {
+	_, gsb := setup(t)
+	if err := gsb.SwitchTo(mockBalancerBuilder1{}); err != nil {
+		t.Fatalf("Balancer.SwitchTo failed with error: %v", err)
+	}
+	if gsb.balancerCurrent == nil {
+		t.Fatal("current balancer not populated after a successful call to SwitchTo()")
+	}
+	// This will be used to update the graceful switch balancer. This update
+	// should simply be forwarded down to the current load balancing policy.
+	ccs := balancer.ClientConnState{
+		BalancerConfig: mockBalancerConfig{},
+	}
+
+	// Updating ClientConnState should forward the update exactly as is to the
+	// current balancer.
+	if err := gsb.UpdateClientConnState(ccs); err != nil {
+		t.Fatalf("Balancer.UpdateClientConnState(%v) failed: %v", ccs, err)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := gsb.balancerCurrent.Balancer.(*mockBalancer).waitForClientConnUpdate(ctx, ccs); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// TestTwoBalancersSameType tests the scenario where there is a graceful switch
+// load balancer setup with a current and pending load balancer of the same
+// type. Any ClientConn update should be forwarded to the current lb if there is
+// a current lb and no pending lb, and the only the pending lb if the graceful
+// switch balancer contains both a current lb and a pending lb. The pending load
+// balancer should also swap into current whenever it updates with a
+// connectivity state other than CONNECTING.
+func (s) TestTwoBalancersSameType(t *testing.T) {
+	tcc, gsb := setup(t)
```

Comment:
This is totally my opinion. You can agree or disagree. But I would ask you to consider it.

When I wrote tests for the RLS LB policy, I decided to write my tests as e2e tests. This means that every test would spin up a server (a stub server exposing the test service), would spin up a stub balancer, would spin up a manual resolver and would do actual RPCs. I found it cumbersome initially (probably for the first test or two, after that my `setup()` and other helper routines were set. so subsequent tests were quick to write). But I'm more happy with the RLS tests now than I have been with any other tests that I have written. I think it gives you immense confidence because you can see it work e2e and tests run super fast because everything is local. 

I feel its worth considering. You won't have to deal with the test ClientConn and wait for updates to propagate etc. You just do an RPC and verify if reaches the backend that you expect. If you are interested, I can walk you through those tests.